### DISCUSSION TITLE
State machines first, React, etc. second

### DISCUSSION THREAD

**Discussion by fubhy** [family/connectkit#252] [General]
Hey there!

I spent a fair amount of time these past few days to evaluate RainbowKit and ConnectKit and pick the best option for my use-case but both fall short in different areas that are impossible to customize and adjust for my (admittedly weird) use case. Essentially, it boils down to both shipping with very opinionated rendering logic deeply tied to the state handling of the connectors.

It's imho not necessary (and limits composability) to encapsulate all that state logic in a waterfall of hooks. Furthermore (and although I'm a React user myself), decoupling it from React would allow others to integrate with Vue or Svelte, etc.

It would be absolutely wonderful if there was a library that merely offered carefully crafted, well tested, fully documented, robust state machines around both: individual connector state and app state (picking a connector, autoconnect, etc.).

Such a design would also massively help with testability ... 

Hence (and this is a long shot), I'd like to propose to completely untangle the state handling & rendering. I think `connectkit` would be a wonderful place to host such an undertaking.

Is this something that you'd be willing to explore?

**Comment by lochie** [comment]
I agree that this would be a good idea, but it would be a bigger project which requires pulling everything apart and rebuilding from the ground up. The UI in particular is pretty core to what ConnectKit provides for a better user experience, though it wouldn't be too crazy to rebuild in vanilla css/js, but to decouple from React would require starting fresh.

Is there a use-case you can share that ConnectKit doesn't cater to? We might be able to help find a solution :)

**Reply by fubhy** [reply]
I mostly wanted to add custom wallets (e.g. impersonator). And because there didn't appear to be an API to register those, I dug through the code a bit to see if I can find a way to fiddle with some internal objects or sth. to inject those. But since there's some two-way coupling going here and there with some of the wrapping components having knowledge of implementation details of specific wallets it didn't seem like that would get me very far either. I just thought it might be a good idea to have a bit less coupling and less leaking of implementation details, but I agree that that would be a bigger undertaking ;-)

**Reply by lochie** [reply]
A way to implement custom wallets currently would be to create a [custom connector](https://github.com/wagmi-dev/references/blob/main/packages/connectors/README.md), or possibly even using the [mock connector](https://wagmi.sh/react/connectors/mock) if you're looking to use impersonator-esque setup.

You can extend the config to include another connector like this: 

```tsx
const ckConfig = getDefaultConfig({
    ...
);

// extend the connectors to include your custom connector / mock connector
ckConfig.connectors = [
  ...ckConfig.connectors,
  myConnector
];

const config = createConfig(ckConfig);

const App = () => {
  return (
    <WagmiConfig config={config}>
      <ConnectKitProvider>
        {/* Your App */}
        <ConnectKitButton />
      </ConnectKitProvider>
    </WagmiConfig>
  );
};
```

**Reply by fubhy** [reply]
Yes but the problem is that I can't add custom UIs there within the connection modal. It's not the end of the world as I can add custom UIs externally that only show if the current connector has a certain `id` but it's not ideal of course from a UX perspective.
